lib/iris/common/metadata.py

```diff
@@ -1339,6 +1342,136 @@ def equal(self, other, lenient=None):
+        :attr:`var_name`. Defaults to value of `default`
+        (which itself defaults to `unknown`) as defined in
+        :class:`~iris.common.CFVariableMixin`.
+
+        (b) a 'coordinate' instance with metadata equal to that of
+        the desired coordinates. Accepts either a
+        :class:`~iris.coords.DimCoord`, :class:`~iris.coords.AuxCoord`,
+        :class:`~iris.aux_factory.AuxCoordFactory`,
+        :class:`~iris.common.CoordMetadata` or
+        :class:`~iris.common.DimCoordMetadata` or
+        :class:`~iris.experimental.ugrid.ConnectivityMetadata`.
+    * standard_name
+        The CF standard name of the desired coordinate. If None, does not
+        check for standard name.
+    * long_name
+        An unconstrained description of the coordinate. If None, does not
+        check for long_name.
+    * var_name
+        The netCDF variable name of the desired coordinate. If None, does
+        not check for var_name.
+    * attributes
+        A dictionary of attributes desired on the coordinates. If None,
+        does not check for attributes.
+    * axis
+        The desired coordinate axis, see
+        :func:`~iris.util.guess_coord_axis`. If None, does not check for
+        axis. Accepts the values 'X', 'Y', 'Z' and 'T' (case-insensitive).
+
+    Returns:
+        A list of the objects supplied in the ``instances`` argument, limited
+        to only those that matched the given criteria.
+
+    """
+    name = None
+    obj = None
+
+    if isinstance(item, str):
+        name = item
+    else:
+        obj = item
+
+    result = instances
+
+    if name is not None:
+        result = [instance for instance in result if instance.name() == name]
+
+    if standard_name is not None:
+        result = [
+            instance
+            for instance in result
+            if instance.standard_name == standard_name
+        ]
+
+    if long_name is not None:
+        result = [
+            instance for instance in result if instance.long_name == long_name
+        ]
+
+    if var_name is not None:
+        result = [
+            instance for instance in result if instance.var_name == var_name
+        ]
+
+    if attributes is not None:
+        if not isinstance(attributes, Mapping):
+            msg = (
+                "The attributes keyword was expecting a dictionary "
+                "type, but got a %s instead." % type(attributes)
+            )
+            raise ValueError(msg)
+
+        def attr_filter(instance):
+            return all(
+                k in instance.attributes
+                and hexdigest(instance.attributes[k]) == hexdigest(v)
+                for k, v in attributes.items()
+            )
+
+        result = [instance for instance in result if attr_filter(instance)]
+
+    if axis is not None:
+        axis = axis.upper()
+        result = [
+            instance
+            for instance in result
+            if guess_coord_axis(instance) == axis
+        ]
```

Comment:
We need to check whether the instance actually has an `axis` attribute, if so use it, otherwise guess it.